# Lab 4 - Information Theory in Machine Learning

Welcome to this week's lab on Information Theory! This week, we will dive into the fascinating world of Information Theory as applied to Machine Learning. Specifically, we will focus on two key concepts: Entropy and Information Gain. These principles are fundamental in understanding how decision trees make split decisions to organize data effectively.

### Entropy
- Entropy, in the context of information theory, measures the level of uncertainty or disorder within a set of data.
- In machine learning, particularly in decision trees, entropy helps to determine how a dataset should be split. A high entropy means more disorder, indicating that our dataset is varied. Conversely, low entropy suggests more uniformity in the data.

### Information Gain
- Information Gain measures the reduction in entropy after the dataset is split on an attribute.
- It is crucial in building decision trees as it helps to decide the order of attributes the tree will use for splitting the data. The attribute with the highest Information Gain is chosen as the splitting attribute at each node.

## Part 1: Entropy and Information Gain in Decision Trees
Decision Trees use these concepts to create branches. By choosing splits that maximize Information Gain (or equivalently minimize entropy), a decision tree can effectively categorize data, leading to better classification or regression models.

### Step 1: Import Necessary Libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree
import matplotlib.pyplot as plt

### Step 2: Load and Explore the Iris Dataset

In [2]:
iris = load_iris()
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df['target'] = iris.target

### Step 3: Calculate Entropy
To calculate the `entropy` we need to:
- First, extract the target variable `y` from your dataset (like the 'target' column in the Iris dataset).
- Then, call `calculate_entropy(y)` to get the entropy.

This function calculates the entropy of a given target variable `y`. It works by first determining the unique classes in `y`, then computes the probability of each class, and uses this probability to calculate the entropy. This is a crucial step in understanding the disorder or uncertainty in the dataset, a fundamental concept in information theory.

In [3]:
def calculate_entropy(y):
    class_labels = np.unique(y)
    entropy = 0
    for label in class_labels:
        probability = len(y[y == label]) / len(y)
        entropy -= probability * np.log2(probability)
    return entropy
print(calculate_entropy(df['target']))
#print(calculate_entropy(df['petal length (cm)']))

1.584962500721156


What is your observastion about the calculated Entropy?

Entropy is based on the portion of each value present in the data. It probably wouldn't be good for continuous data and is probably best off used on labeled data.

### Step 4: Calculate Information Gain
There are three steps for calculating the Information Gain:
1. Compute Overall Entropy: Use the entropy function from Step 3 on the entire target dataset.
2. Calculate Weighted Entropy for Each Attribute: For each unique value in the attribute, partition the dataset and calculate its entropy. Then calculate the weighted sum of these entropies, where the weights are the proportions of instances in each partition.
3. Compute Information Gain: Subtract the weighted entropy of the split from the original entropy.

The attribute with the highest Information Gain is generally chosen for splitting, as it provides the most significant reduction in uncertainty. This step is critical in constructing an effective decision tree, as it directly influences the structure and depth of the tree.

In [4]:
def calculate_information_gain(df, attribute, target_name):
    total_entropy = calculate_entropy(df[target_name])
    values, counts = np.unique(df[attribute], return_counts=True)
    weighted_entropy = sum((counts[i] / sum(counts)) * calculate_entropy(df.where(df[attribute] == values[i]).dropna()[target_name]) for i in range(len(values)))
    information_gain = total_entropy - weighted_entropy
    return information_gain


print(calculate_information_gain(df, 'petal width (cm)', 'target'), calculate_information_gain(df, 'petal length (cm)', 'target'), calculate_information_gain(df, 'sepal width (cm)', 'target'), calculate_information_gain(df, 'sepal length (cm)', 'target'))

1.4358978386754417 1.4463165236458 0.5166428756804977 0.8769376208910578


Discuss your findings here.
It seems like the most important one is the petal width, and the least important one is the sepal width.


## Part 2: Apply Entropy and Information Gain on a different dataset

Your task is to choose a new dataset and implement what you learned in `Part 1` on this new dataset.

### Task 1: Implement Entropy and Information Gain

In [5]:
from sklearn.datasets import load_wine
wine =load_wine()
# Your code goes here
wine
df2=pd.DataFrame(wine.data, columns=wine.feature_names)
df2['target']=wine.target
#df2['target'].value_counts()

#print(wine.feature_names)

entropy2=calculate_entropy(df2['target'])
print("entropy of wine dataset:", entropy2)

for i in wine.feature_names:
    print (i, calculate_information_gain(df2, i, 'target'))

entropy of wine dataset: 1.5668222768551812
alcohol 1.3406163077175945
malic_acid 1.2842014538975586
ash 0.7863398170147985
alcalinity_of_ash 0.7357938479915789
magnesium 0.6361513778259005
total_phenols 1.1812650269501774
flavanoids 1.4207548611248442
nonflavanoid_phenols 0.4739066456615457
proanthocyanins 1.1161486280647779
color_intensity 1.3700832318627436
hue 1.0413965047465832
od280/od315_of_diluted_wines 1.3470502797834747
proline 1.3015454175918462


### Task 2: Discuss your findings in detail

The overall entropy of the wine dataset is around 1.567. The most important category is the falavanoids, and the least important category is nonflavanoid phenols.

## Submission
Submit a link to your completed Jupyter Notebook file hosted on your private GitHub repository through the submission link in Blackboard.